### PR TITLE
Cluster certificate authority rotation, implements #1860

### PR DESCRIPTION
**Pull request by klizhentas**
This PR implements #1860

```
    Teleport cert authority rotation.
    
    During the the rotation procedure issuing TLS and SSH
    certificate authorities are re-generated and all internal
    components of the cluster re-register with the new
    credentials.
    
    The rotation procedure is based on the distributed
    state machine algorithm - certificate authorities have
    explicit rotation state and all parts of the cluster sync
    local state machines following transitions between phases.
    
    Operator can launch CA rotation in auto or manual modes.
    
    In manual mode operator moves bewtween rotation states
    and watches the states of the components to sync.
    
    In auto mode state transitions are happening automatically
    on a specific schedule.
    
    The design documentation is embedded in the code:
    
    lib/auth/rotate.go
```

Teleport.e counterpart:

https://github.com/gravitational/teleport.e/pull/69

### CODE REVIEW

**Comment by klizhentas** [comment]
retest this please

**Comment by klizhentas** [comment]
retest this please

File: lib/backend/dir/impl.go [hunk 1]
```diff
@@ -174,6 +175,57 @@ func (bk *Backend) CreateVal(bucket []string, key string, val []byte, ttl time.D
...
+		return trace.Wrap(err)
+	}
+	defer utils.FSUnlock(f)
+	// before writing, make sure the values are equal
+	oldVal, err := ioutil.ReadAll(f)
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	if bytes.Compare(oldVal, prevVal) != 0 {
+		return trace.CompareFailed("%v/%v did not match expected value", dirPath, key)
```
Comment (r0mant) [inline lib/backend/dir/impl.go]: Maybe add "expected value" to the message as well?

File: lib/services/authority.go [hunk 2]
```diff
@@ -520,11 +560,198 @@ func (ca *CertAuthorityV2) CheckAndSetDefaults() error {
...
+			"unsupported rotation 'state': %q, supported states are: %q, %q",
+			r.State, RotationStateStandby, RotationStateInProgress)
+	}
+	return nil
+}
+
+// GenerateSchedule generates schedule based on the time period, using
+// even time periods between rotation phases.
+func GenerateSchedule(clock clockwork.Clock, gracePeriod time.Duration) (*RotationSchedule, error) {
+	if gracePeriod == 0 {
```
Comment (r0mant) [inline lib/services/authority.go]: Check for `< 0` too?

File: integration/helpers.go [hunk 3]
```diff
@@ -832,7 +953,7 @@ func startAndWait(process *service.TeleportProcess, expectedEvents []string) ([]
 	// register to listen for all ready events on the broadcast channel
 	broadcastCh := make(chan service.Event)
 	for _, eventName := range expectedEvents {
-		process.WaitForEvent(eventName, broadcastCh, make(chan struct{}))
+		process.WaitForEvent(context.TODO(), eventName, broadcastCh)
```
Comment (r0mant) [inline integration/helpers.go]: Would it make sense for the caller to pass context to "startAndWait?

File: lib/auth/rotate.go [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+
+	rotation.Mode = req.mode
+	rotation.Schedule = req.schedule
+
+	var sshPrivPEM, sshPubPEM []byte
+	var keyPEM, certPEM []byte
+
+	// generate keys and certificates:
+	if len(req.privateKey) != 0 {
+		log.Infof("Generating CA in test mode.")
```
[the remark this answers is not in the document]
Comment (klizhentas) [reply lib/auth/rotate.go]: yes, updated

File: lib/auth/rotate.go [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+		signingKeys = [][]byte{sshPrivPEM}
+		checkingKeys = [][]byte{sshPubPEM}
+		keyPairs = []services.TLSKeyPair{*tlsKeyPair}
+		// In case of forced rotation, rotation has been started and completed
+		// in the same step moving it to standby state.
+		rotation.State = services.RotationStateStandby
+	} else {
+		// Rotation sets the first key to be the new key
+		// and keep only public keys/certs for the new CA.
+		signingKeys = [][]byte{sshPrivPEM, signingKeys[0]}
```
[the remark this answers is not in the document]
Comment (klizhentas) [reply lib/auth/rotate.go]: the checks are performed one level up, to make sure these functions are called from certain states, so if there are not enough element in the list, the data is corrupted, so panic will be appropriate.

File: lib/auth/rotate.go [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+	keyPairs := ca.GetTLSKeyPairs()
+
+	// Drop old certificate authority without keeping it as trusted.
+	if gracePeriod == 0 {
+		signingKeys = [][]byte{sshPrivPEM}
+		checkingKeys = [][]byte{sshPubPEM}
+		keyPairs = []services.TLSKeyPair{*tlsKeyPair}
+		// In case of forced rotation, rotation has been started and completed
+		// in the same step moving it to standby state.
+		rotation.State = services.RotationStateStandby
```
[the remark this answers is not in the document]
Comment (klizhentas) [reply lib/auth/rotate.go]: same thing, but will update for consistency.

File: lib/backend/dir/impl.go (hunk 1 above)
Comment (klizhentas) [reply lib/backend/dir/impl.go]: the expected value could be a large JSON object, so it will make a very hard to read error message.

**Comment by klizhentas** [comment]
retest this please

**Review by r0mant — APPROVED** [review]

**Review by russjones — APPROVED** [review]

**Pull request merged**